src/layout/node.ts

```diff
@@ -17,29 +17,44 @@ export default function(node: Node, source: Source): void {
     }
 
     let call = dom.has(node) ? "update" : "add";
+    let parent = node.parentElement ? node.parentElement : null;
     switch (node.nodeType) {
         case Node.DOCUMENT_TYPE_NODE:
             let doctype = node as DocumentType;
             let docAttributes = { name: doctype.name, publicId: doctype.publicId, systemId: doctype.systemId };
             let docData = { tag: "*D", attributes: docAttributes };
-            dom[call](node, docData, source);
+            dom[call](node, parent, docData, source);
+            break;
+        case Node.DOCUMENT_FRAGMENT_NODE:
+            let shadowRoot = (node as ShadowRoot);
```

Comment:
Is there any chance this might not be a Shadow DOM node and the shadowRoot would be null? Might be worth changing condition below to if(shadowRoot && )
Also consider updating to TypeScript 3.7 as it finally has option chanhing (.?) to avoid all the if(x && x.foo) or x == null ? x.foo : null junk.
See: [https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html)

Reply:
shadowRoot is just a typed version of node for Typescript. So if node is null, we wouldn't be able to get inside the switch / case anyway because the uber switch condition is node.nodeType.